src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/KubernetesCredentialProvider.java

```diff
@@ -72,40 +78,54 @@
 
     private KubernetesCredentialsStore store = new KubernetesCredentialsStore(this);
 
+    KubernetesClient getKubernetesClient() {
+        if (client == null) {
+            ConfigBuilder cb = new ConfigBuilder();
+            Config config = cb.build();
+            client = new DefaultKubernetesClient(config);
+        }
+        return client;
+    }
+
     @Initializer(after=InitMilestone.PLUGINS_PREPARED, fatal=false)
     @Restricted(NoExternalUse.class) // only for callbacks from Jenkins
     public void startWatchingForSecrets() {
+        final String initAdminMonitorId = getClass().getName() + ".initialize";
         try {
-            ConfigBuilder cb = new ConfigBuilder();
-            Config config = cb.build();
-            DefaultKubernetesClient _client = new DefaultKubernetesClient(config);
-            LOG.log(Level.FINER, "Using namespace: {0}", _client.getNamespace());
-            LOG.log(Level.FINER, "retreiving secrets");
-            SecretList list = _client.secrets().withLabel(SecretUtils.JENKINS_IO_CREDENTIALS_TYPE_LABEL).list();
+            KubernetesClient _client = getKubernetesClient();
+            LOG.log(Level.FINER, "Using namespace: {0}", String.valueOf(_client.getNamespace()));
+
+            // start watching new secrets before we list the current set of secrets so we don't miss any events
+            LOG.log(Level.FINER, "registering watch");
+            watch = _client.secrets().withLabel(SecretUtils.JENKINS_IO_CREDENTIALS_TYPE_LABEL).watch(this);
+            LOG.log(Level.FINER, "registered watch, retrieving secrets");
 
+            // load current set of secrets into provider
+            LOG.log(Level.FINER, "retrieving secrets");
+            SecretList list = _client.secrets().withLabel(SecretUtils.JENKINS_IO_CREDENTIALS_TYPE_LABEL).list();
+            // reset credentials to the latest full listing
+            credentials.clear();
```

Comment:
anyone asking for credentials at this point will get nothing back (for a transient issue).  that was why the credentials where not cleared last time but the reference was changed (such that we always have a fully populated list (even if it is stale, its never half populated))